pkg/liqonet/netns/netns.go

```diff
@@ -0,0 +1,78 @@
+package netns
+
+import (
+	"runtime"
+
+	"github.com/containernetworking/plugins/pkg/ip"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+	"github.com/vishvananda/netns"
+	"golang.org/x/sys/unix"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+
+	"github.com/liqotech/liqo/pkg/liqonet"
+)
+
+const (
+	nsPath = "/run/netns/"
+)
+
+// CreateNetns given a name it will check if a namespace exists with the given name
+// and will remove it. Then the namespace will be recreated.
+// Returns a handler to the newly created network namespace or an error in case
+// something goes wrong.
+func CreateNetns(name string) (ns.NetNS, error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+	namespacePath := nsPath + name
+	err := DeleteNetns(name)
+	if err != nil {
+		return nil, err
+	}
+	// Create a new network namespace.
+	_, err = netns.NewNamed(name)
+	if err != nil {
+		return nil, err
+	}
+	netNamespace, err := ns.GetNS(namespacePath)
+	if err != nil {
+		return nil, err
+	}
+	return netNamespace, nil
+}
+
+// DeleteNetns removes a given network namespace by name.
+// If the namespace does not exist does nothing, in case of error returns it.
+func DeleteNetns(name string) error {
+	if err := netns.DeleteNamed(name); err != nil && !errors.IsError(err, unix.ENOENT) {
+		klog.Errorf("an error occurred while removing network namespace with name %s: %v", name, err)
+		return err
+	}
+	return nil
+}
+
+// CreateVethPair it will create veth pair in originNetns and move one of them in dstNetns.
+// Error is returned if something goes wrong.
+func CreateVethPair(originVethName, dstVethName string, originNetns, dstNetns ns.NetNS, linkMTU int) error {
+	if originNetns == nil || dstNetns == nil {
+		return &liqonet.WrongParameter{
+			Parameter: "originNetns and dstNetns",
+			Reason:    liqonet.NotNil}
+	}
+	var createVethPair = func(hostNS ns.NetNS) error {
+		_, _, err := ip.SetupVethWithName(originVethName, dstVethName, linkMTU, dstNetns)
+		if err != nil {
+			klog.Errorf("an error occurred while creating veth pair between host and gateway namespace: %v", err)
+			return err
+		}
+		return nil
+	}
+	retryiable := func(err error) bool {
```

Comment:
Why do you need a retriable here?

Reply:
The veth pair has one one device on the host network and the other one on the custom namespace. Some times if we delete the custom namespace it wold take some time for the kernel to remove the veth device on the host network, and a conflict could arise while reinstalling the veth device in the host network. I also added a comment.